src/Interpreters/CrossToInnerJoinVisitor.cpp

```diff
@@ -80,147 +81,150 @@ struct JoinedElement
-                 functionIsInOperator(node.name))  /// IN, NOT IN
-        {
-            /// leave as is. It's not possible to make push down here cause of unknown aliases and not implemented JOIN predicates.
-            ///     select a as b form t1, t2 where t1.x = t2.x and b in(42)
-            ///     select a as b form t1 inner join t2 on t1.x = t2.x and b in(42)
-        }
-        else
-        {
-            ands_only = false;
-            asts_to_join_on.clear();
-        }
+    static void visit(const ASTPtr & node, Data & data)
+    {
+        if (const auto * ident = node->as<ASTIdentifier>())
+            data.idents.push_back(ident);
     }
 
-    bool complex() const { return !ands_only; }
-    bool matchAny(size_t t) const { return asts_to_join_on.count(t); }
-
-    ASTPtr makeOnExpression(size_t table_pos)
+    static bool needChildVisit(const ASTPtr &, const ASTPtr &)
     {
-        if (!asts_to_join_on.count(table_pos))
-            return {};
-
-        std::vector<ASTPtr> & expressions = asts_to_join_on[table_pos];
-
-        if (expressions.size() == 1)
-            return expressions[0]->clone();
+        return true;
+    }
 
-        std::vector<ASTPtr> arguments;
-        arguments.reserve(expressions.size());
-        for (auto & ast : expressions)
-            arguments.emplace_back(ast->clone());
+    static ASTIdentifiers collect(const ASTPtr & node)
+    {
+        IdentifiersCollector::Data ident_data;
+        ConstInDepthNodeVisitor<IdentifiersCollector, true> ident_visitor(ident_data);
+        ident_visitor.visit(node);
+        return ident_data.idents;
+    }
+};
 
-        return makeASTFunction(NameAnd::name, std::move(arguments));
+/// Split expression `expr_1 AND expr_2 AND ... AND expr_n` into vector `[expr_1, expr_2, ..., expr_n]`
+void collectConjunctions(const ASTPtr & node, std::vector<ASTPtr> & members)
+{
+    if (const auto * func = node->as<ASTFunction>(); func && func->name == NameAnd::name)
+    {
+        for (const auto & child : func->arguments->children)
+            collectConjunctions(child, members);
+        return;
     }
+    members.push_back(node);
+}
 
-private:
-    const std::vector<JoinedElement> & joined_tables;
-    const std::vector<TableWithColumnNamesAndTypes> & tables;
-    std::map<size_t, std::vector<ASTPtr>> asts_to_join_on;
-    const Aliases & aliases;
-    bool ands_only;
+std::vector<ASTPtr> collectConjunctions(const ASTPtr & node)
+{
+    std::vector<ASTPtr> members;
+    collectConjunctions(node, members);
+    return members;
+}
 
-    size_t canMoveEqualsToJoinOn(const ASTFunction & node)
-    {
-        if (!node.arguments)
-            throw Exception("Logical error: function requires arguments", ErrorCodes::LOGICAL_ERROR);
-        if (node.arguments->children.size() != 2)
-            return false;
+std::optional<size_t> getIdentMembership(const ASTIdentifier & ident, const std::vector<TableWithColumnNamesAndTypes> & tables)
```

Comment:
Nice, this almost starts to look like the proper join clause infrastructure. Eventually we should have an explicit data structure describing the join tree, and then for each WHERE clause, check which relations participate in it, and put it to the appropriate level in this tree.

Reply:
I also tried to use this code in https://github.com/ClickHouse/ClickHouse/pull/21370 to fix problem with `Merge` engine. While executing JOIN with `Merge` there are step when we remove right table from query and perform analysis on it (I don't completly figure out why it is needed, maybe it requires other approach). But expression that uses right table left in query and it leads to error. This code can be used to remove them too.